### ISSUE TITLE
Different behavior in Chrome vs Firefox

### ISSUE THREAD

**Issue by smfelsher** [TonyBrobston/jpegasus#141]
Using your demo page, https://tonybrobston.github.io/jpegasus-demo/, I see a difference between Firefox and Chrome when uploading a GIF file. Chrome produces a smaller JPEG, but Firefox produces a larger JPEG.

**Comment by TonyBrobston** [comment]
@smfelsher I've found that different browsers or even the same browser but on a different operating system produces different sized outputs. 

I haven't really looked into why, but my assumption is that there is some sort of spec that a browser has to adhere to when implementing certain features, like `HTMLCanvasElement`, `HTMLImageElement`, etc.. I assume this is similar to an interface, so like the element name, functions, parameters of functions, etc. likely have to match the spec. But, that does not mean the internal implementation of those elements are the same, thus producing a different result while still adhering to the spec/contract.

Again, this is an assumption on my part. The intent of this library is to take in a JavaScript File along with some options and return a JavaScript File, just smaller. It does that, so I'm not sure that there is a problem. Yes, ideally it would return the same size, but it doesn't 🤷‍♂️ .

Thanks for the feedback though, if you have any other feedback I'm definitely open to trying to solve it.

**Comment by smfelsher** [comment]
Thank you @TonyBrobston for the detailed explanation. Your point about 

> The intent of this library is to take in a JavaScript File along with some options and return a JavaScript File, just smaller.

made me realize that I did not clearly explain what I was seeing.

I visit your demo page in the Firefox browser and select a GIF that is 11.09 Kbytes in size. When I press submit, the tool indicates that `jpegasus` produces a JPEG that is 37.04 Kbytes. This result is __larger__ than the original.

When I visit the demo page with the Chrome browser and perform the exact same steps, the result is a __smaller__ file.

I am using `Cypress` to test my application and was incurring a failure for my file upload tests when the environment was Firefox. I see that you have an open issue to test `jpegasus` in multiple browsers with `Cypress`, so maybe this is a good test scenario.

**Comment by TonyBrobston** [comment]
@smfelsher I think maybe talking from the perspective of "what problem are we trying to solve?", might be helpful.

I think the problem is, `jpegasus` should always return a file the same size or smaller.

I'd be open to adding an optional boolean of `returnOriginalIfBigger` (or something along those lines).

The underlying technology is `HTMLCanvasElements`' `toDataURL` function. In my options `quality` is fed directly to `encoderOptions` which is documented as: "A Number between 0 and 1 indicating the image quality to use for image formats that use lossy compression such as image/jpeg and image/webp.
If this argument is anything else, the default value for image quality is used. The default value is 0.92. Other arguments are ignored."

https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toDataURL

What I've found is, if I put the attached image into https://tonybrobston.github.io/jpegasus-demo/ with `maxHeight` 3000, `maxWidth` 4000, and then play around with quality 0.92 thru 1.00 I get interesting results; which I was previously aware of. 0.95 produces close to the original quality, when really I expect 1.00 to produce exactly the original quality. (I realize that the original image was 4032 x 3024, so the 32 x 24 less area will account for some of that size loss; maybe my demo ui should allow exact selection of height/width or an option for `null` or `original` or something like that.) 
![IMG_20200208_132400](https://user-images.githubusercontent.com/4724577/81956858-6e552a00-95d1-11ea-9582-e995e315f89d.jpg)

So I sort of blame `HTMLCanvasElements`' `toDataURL` for the confusion. I could scale `jpegasus`' `qualtiy` option of 1.00 to equal `encoderOptions` 0.95; but I'm not sure if that adds a lot of value. Some day in the future I'd really like to implement an options of `targetFileSize`, as I think the real feature I (and I think others) would want is to take an image bigger than X and compress it to exactly X. So maybe a site limits you to 2mb files, but you took a picture with your smart phone and it's 5mb, ideally you'd lose the least amount of quality. However this is a pretty complicated thing to solve. I'm not sure of the best way to solve it, but I've planned to play around with neural network training, which I think could solve it given enough training data (https://www.npmjs.com/package/brain.js).

Let me know what you think of `returnOriginalIfBigger` or if you have another idea of how to solve the problem or another problem to describe.

I'm hoping this conversation is helpful, it certainly is helpful for me to talk to someone else that is engaged with the problem and how to solve it.

**Comment by smfelsher** [comment]
I'm coming at this from the 50,000 ft view. I've only been "web" programming for about 6 months (embedded programming for about 20 years), so I only have a simplistic view of the problem :).

What I have done in my application is exactly what you suggest, If `jpegagus` returns a file that is larger than the original, then I save the original to my server; otherwise, I'll use the smaller file that `jpegasus` returns.

I can give you some background on how I'm using your library. My end-users are recording "incidents" in a database. They want to also be able to attach files: mostly images, but I should expect "anything". I'm storing the "attachments" on a file server rather than in the database. They don't want to burn through hard-disk space with images, so I am using your tool to "compress" the images. For any mime type that is `image`, I call you library with:
```js
{
  maxHeight: 1000,
  maxWidth: 1000,
  quality: 0.65,
  scaleImageBy: 1,
  returnOriginalOnFailure: true
}
```
These options produced images that were more than acceptable when I manually tested on your demo site with some JPEGs that my client sent me.

What I discovered during integration testing (E2E with Cypress) when I was sending different image types was that Firefox was giving a different result than Chrome for a GIF (I test with a JPEG, PNG, and a GIF, all are `2000 x 2000`). When your library receives a GIF that is `2000 x 2000`, it produces a JPEG that is `1000 x 1000`, but the file size is larger than the original. So, that prompted me to:

1. Add the condition that I store the smaller file size between the original and the result of compression
1. Expect a different result when I E2E test with the Firefox browser vs Chrome browser.

From my standpoint, I was not expecting the browser to change the results produced by your library when the input and configuration are exactly the same for both browsers. 

_So, my question is "Do you expect your library to produce the same results given the exact same inputs but two (2) different browsers?"_

I think the `returnOriginalIfBigger` option would come in useful in my case **_and_** you think that is expected behavior.

If the conversation is helpful for you, then let's continue it.

**Comment by TonyBrobston** [comment]
I've done a decent amount of frontend and api development (close to 9 years between the two), I have no experience with embedded; but logic is logic, so I imagine with 20 years of programming you'll pick this stuff up quick.

The background definitely helps. It's also worth mentioning that compressing client side also means you're sending a smaller file, so less bandwidth as well as faster uploads.

A few other things worth mentioning:
`scaleBy` defaults to `1`, so you can remove that from your options if you want.
I do validate the file before trying to compress it:
https://github.com/TonyBrobston/jpegasus/blob/6f4861020c3ac59f05e384a804bb5fc2dced9313/src/index.ts#L11
https://github.com/TonyBrobston/jpegasus/blob/6f4861020c3ac59f05e384a804bb5fc2dced9313/src/services/fileService.ts#L1

As of right now I don't expect two different browsers with the same inputs to return the same output; it would be ideal if they did, but they don't. Again this is mainly because the browser itself is doing the heavy lifting and I'm abstracted away from that.

I've begun implementation of a "return if bigger", I think I'm going to call it `returnOriginalIfCompressedIsLarger` or `returnOriginalIfCompressedFileIsLarger`.

I'll have something implemented shortly, then we can discuss if there are any other needs.

**Comment by smfelsher** [comment]
Why do you think that Firefox is producing a larger file size when it converts the GIF to JPEG?

**Comment by TonyBrobston** [comment]
Here are the changes I've put together.
https://github.com/TonyBrobston/jpegasus/pull/149

Can you provide the gif you're referring to? I grabbed a gif randomly off the internet and I found that with `maxHeight` 1000, `maxWidth` 1000, quality `1`. Original size 3003.4 kb, compressed Chrome 114.82 kb, compressed Firefox 120.67 kb. Are you asking why I think Firefox is 5.85 kb bigger?

**Comment by smfelsher** [comment]
I'm not asking why Firefox is 5.85 kb bigger than Chrome. 

I'm asking why `jpegasus` is not compressing my GIF file, but actually making it larger, when I use the Firefox browser.

When I upload the attached image to the `jpegasus` demo using the Firefox browser, the demo page reports:

```
Original    11.09 kb   image/gif
Jpegasus   37.04 kb image/jpeg
```
1. Why is `jpegaus` not compressing the file?
2. When I do this with the Chrome browser, the **Jpegasus** file is smaller than the original.

![2000](https://user-images.githubusercontent.com/25533834/81993329-73d36400-9613-11ea-9f52-d454f459be9c.gif)

**Comment by smfelsher** [comment]
OK, now when I try the demo page with the Chrome browser, it behaves like Firefox. The **Jpegasus** file size is larger than the original. This morning, this was not the case.

**Comment by smfelsher** [comment]
Sorry, I didn't configure things properly.
```
Max Height: 1000
Max Width: 1000
Scale Image By: 1
Quality: 0.65
Upload Image: 2000.gif
```

Firefox
```
Original 11.09 kb image/gif
Jpegasus 18.36 kb image/jpeg
```
Chrome 
```
Original 11.09 kb image/gif
Jpegasus 8.69  kb image/jpeg
```
Chrome compressed, Firefox made it **larger**.

**Comment by TonyBrobston** [comment]
I still don't have a solid answer as to why Firefox ends up larger, but I do have `returnOriginalIfCompressedFileIsLarger` implemented.

I have the demo site setup as a Progressive Web Application and I don't have a way setup yet to automatically unregister the old service worker. So you'll either need to open developer tools and go to Application > Service Workers and click "Unregister" then refresh or open an Incognito window. Sorry for the inconvenience on that, I'll see about setting it up to happen automatically.

https://tonybrobston.github.io/jpegasus-demo/

**Comment by smfelsher** [comment]
I'll update the library and utilize the new configuration option ASAP.

Thanks!
